demo/malloy-demo-composer/src/app/Explore/Explore.tsx

```diff
@@ -11,393 +11,112 @@
-  const queryBuilder = useRef<QueryBuilder>();
-  const [queryMalloy, setQueryMalloy] = useState<string>("");
-  const [querySummary, setQuerySummary] = useState<QuerySummary>();
-  const [queryName, setQueryName] = useState("");
-  const [error, setError] = useState<Error>();
   const {
-    result,
-    runQuery: runQueryRaw,
+    queryMalloy,
+    queryName,
+    clearQuery,
+    runQuery,
     isRunning,
     clearResult,
-  } = useRunQuery(queryMalloy, queryName, setError, analysis);
-  const { saveField } = useSaveField(openDirectory, analysis, (newAnalysis) => {
-    setAnalysis(newAnalysis);
-    withAnalysisSource(newAnalysis, (source) => {
-      queryBuilder.current?.updateSource(source);
-    });
-  });
-  const [insertOpen, setInsertOpen] = useState(false);
-  const [dataStyles, setDataStyles] = useState<DataStyles>({});
-  const topValues = useTopValues(analysis);
-
-  useWatchAnalysis(analysis, (newAnalysis) => {
-    setAnalysis(newAnalysis);
-    withAnalysisSource(newAnalysis, (source) => {
-      queryBuilder.current?.updateSource(source);
-    });
+    source,
+    queryModifiers,
+    querySummary,
+    dataStyles,
+    result,
+    error,
+  } = useQueryBuilder({
+    analysis,
+    setAnalysis,
+    openDirectory,
   });
 
-  const selectAnalysis = (analysis: Analysis) => {
-    setAnalysis(analysis);
-    clearQuery(analysis);
-  };
-
-  const runQuery = () => {
-    runQueryRaw();
-    const topLevel = {
-      stageIndex: querySummary ? querySummary.stages.length - 1 : 0,
-    };
-    if (!queryBuilder.current?.hasLimit(topLevel)) {
-      // TODO magic number here: we run the query before we set this limit,
-      //      and this limit just happens to be the default limit
-      addLimit(topLevel, 10);
-    }
-  };
+  const topValues = useTopValues(analysis);
+  const [section, setSection] = useState("query");
 
-  const writeQuery = (newDataStyles = dataStyles, newAnalysis = analysis) => {
-    if (!queryBuilder.current) {
-      return;
-    }
-    const query = queryBuilder.current.getQuery();
-    setQueryName(query.name);
-    // eslint-disable-next-line no-console
-    console.log(query);
-    if (!newAnalysis) {
-      return;
-    }
-    withAnalysisSource(newAnalysis, (source) => {
-      const writer = new QueryWriter(query, source);
-      if (queryBuilder.current?.canRun()) {
-        const queryString = writer.getQueryStringForModel();
-        setQueryMalloy(queryString);
-        // eslint-disable-next-line no-console
-        console.log(queryString);
+  const loadQueryLink = (
+    modelPath: string,
+    sourceName: string,
+    queryName: string
+  ) => {
+    let current: Directory | Model | Analysis | undefined = directory;
+    // Note, this only works for relative paths like ./dir/model.malloy
+    // and cannot go up the directory hierarchy. Therefore, it will only load
+    // models in the same directory.
+    for (const segment of modelPath.split("/")) {
```

Comment:
I'm not really sure of a better way to handle this for now, given that we're loading the whole directory and then just switching between the models locally. This sorta means that a project can only link to models in the same project, because nothing else is loaded. This seems like an ok limitation.